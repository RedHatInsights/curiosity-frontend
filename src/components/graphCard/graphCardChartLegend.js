import React from 'react';
import PropTypes from 'prop-types';
import { Button, Tooltip, TooltipPosition } from '@patternfly/react-core';
import { EyeSlashIcon } from '@patternfly/react-icons';
import { connectTranslate, store, reduxTypes } from '../../redux';
import { helpers } from '../../common';

/**
 * A custom chart legend.
 *
 * @augments React.Component
 * @fires onClick
 */
class GraphCardChartLegend extends React.Component {
  componentDidMount() {
    const { chart, datum, legend, viewId } = this.props;
    datum.dataSets.forEach(({ id }) => {
      const checkIsToggled = legend[`${viewId}-${id}`] || chart.isToggled(id);

      if (checkIsToggled) {
        chart.hide(id);
      }
    });
  }

  /**
   * Toggle legend item and chart.
   *
   * @event onClick
   * @param {string} id
   */
  onClick = id => {
    const { chart, viewId } = this.props;
    const updatedToggle = chart.toggle(id);

    store.dispatch({
      type: reduxTypes.graph.SET_GRAPH_LEGEND,
      legend: {
        [`${viewId}-${id}`]: updatedToggle
      }
    });
  };

  /**
   * Return a legend item.
   *
   * @param {object} options
   * @param {string} options.chartId
   * @param {string} options.color
   * @param {boolean} options.isDisabled
   * @param {boolean} options.isThreshold
   * @param {string} options.labelContent
   * @param {string} options.tooltipContent
   * @returns {Node}
   */
  renderLegendItem({ chartId, color, isDisabled, isThreshold, labelContent, tooltipContent }) {
    const { chart, legend, viewId } = this.props;
    const checkIsToggled = legend[`${viewId}-${chartId}`] || chart.isToggled(chartId);

    const button = (
      <Button
        onClick={() => this.onClick(chartId)}
        onKeyPress={() => this.onClick(chartId)}
        className="victory-legend-item"
        tabIndex={0}
        key={`curiosity-button-${chartId}`}
        variant="link"
        component="a"
        isDisabled={isDisabled}
        icon={
          ((isDisabled || checkIsToggled) && <EyeSlashIcon />) ||
          (isThreshold && (
            <hr
              aria-hidden
              className="threshold-legend-icon"
              style={{
                visibility: (isDisabled && 'hidden') || (checkIsToggled && 'hidden') || 'visible',
                borderTopColor: color
              }}
            />
          )) || (
            <div
              aria-hidden
              className="legend-icon"
              style={{
                visibility: (isDisabled && 'hidden') || (checkIsToggled && 'hidden') || 'visible',
                backgroundColor: color
              }}
            />
          )
        }
      >
        {labelContent}
      </Button>
    );

    if (tooltipContent) {
      return (
        <Tooltip
          key={`curiosity-tooltip-${chartId}`}
          content={<p>{tooltipContent}</p>}
          position={TooltipPosition.top}
          distance={-10}
          entryDelay={100}
          exitDelay={0}
        >
          {button}
        </Tooltip>
      );
    }

    return button;
  }

  /**
   * Render a graph legend.
   *
   * @returns {Node}
   */
  render() {
    const { datum, product, t } = this.props;

    return (
      <React.Fragment>
        {datum.dataSets.map(({ id, isThreshold, stroke, data = [] }) => {
          const isDisabled =
            !data.find(({ y, hasData }) => (y >= 0 && hasData === true) || (y >= 0 && isThreshold === true)) || false;

          const labelContent =
            (isThreshold && t(`curiosity-graph.thresholdLabel`)) ||
            t(`curiosity-graph.${id}Label`, { product }) ||
            t(`curiosity-graph.noLabel`, { product });

          // ToDo: Disabled tooltip content, reactivate accordingly, issues/158
          // const tooltipContent =
          //  (isThreshold && t(`curiosity-graph.thresholdLegendTooltip`)) || t(`curiosity-graph.${id}LegendTooltip`);

          return this.renderLegendItem({
            chartId: id,
            color: stroke,
            labelContent,
            isDisabled,
            isThreshold
            // tooltipContent
          });
        })}
      </React.Fragment>
    );
  }
}

/**
 * Prop types.
 *
 * @type {{datum, product: string, t: Function, legend, chart}}
 */
GraphCardChartLegend.propTypes = {
  chart: PropTypes.shape({
    hide: PropTypes.func,
    toggle: PropTypes.func,
    isToggled: PropTypes.func
  }),
  datum: PropTypes.shape({
    dataSets: PropTypes.arrayOf(
      PropTypes.shape({
        stroke: PropTypes.string.isRequired,
        id: PropTypes.string.isRequired,
        isThreshold: PropTypes.bool
      })
    )
  }),
  legend: PropTypes.objectOf(PropTypes.bool),
  product: PropTypes.string,
  t: PropTypes.func,
  viewId: PropTypes.string
};

/**
 * Default props.
 *
 * @type {{datum: { dataSets: [] }, product: string, t: Function, legend: {}, chart: {hide: Function,
 *     toggle: Function, isToggled: Function}}}
 */
GraphCardChartLegend.defaultProps = {
  chart: {
    hide: helpers.noop,
    toggle: helpers.noop,
    isToggled: helpers.noop
  },
  datum: {
    dataSets: []
  },
  legend: {},
  product: '',
  t: helpers.noopTranslate,
  viewId: 'graphCardLegend'
};

const mapStateToProps = ({ graph }) => ({ legend: graph.legend });

const ConnectedGraphCardChartLegend = connectTranslate(mapStateToProps)(GraphCardChartLegend);

export { ConnectedGraphCardChartLegend as default, ConnectedGraphCardChartLegend, GraphCardChartLegend };
