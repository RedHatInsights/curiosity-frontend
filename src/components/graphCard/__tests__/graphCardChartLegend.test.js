import React from 'react';
import { shallow } from 'enzyme';
import { Button } from '@patternfly/react-core';
import { GraphCardChartLegend } from '../graphCardChartLegend';

describe('GraphCardChartLegend Component', () => {
  it('should render a basic component', () => {
    const props = {
      datum: {
        dataSets: [
          {
            stroke: '#000000',
            id: 'loremIpsum',
            data: [{ y: 0, hasData: true }]
          }
        ]
      }
    };
    const component = shallow(<GraphCardChartLegend {...props} />);

    expect(component).toMatchSnapshot('basic');
  });

  it('should render basic data', () => {
    const props = {
      datum: {
        dataSets: [
          {
            stroke: '#000000',
            id: 'loremIpsum',
            isThreshold: false,
            data: [{ y: 0, hasData: true }]
          },
          {
            stroke: '#000000',
            id: 'ametConsectetur',
            isThreshold: false,
            data: [{ y: 0, hasData: false }]
          },
          {
            stroke: '#ff0000',
            id: 'dolorSit',
            isThreshold: true,
            data: [{ y: 0, isInfinite: false }]
          },
          {
            stroke: '#ff0000',
            id: 'nonCursus',
            isThreshold: true,
            data: [{ y: 0, isInfinite: true }]
          }
        ]
      },
      productLabel: 'test'
    };

    const component = shallow(<GraphCardChartLegend {...props} />);
    expect(component).toMatchSnapshot('data');
  });

  it('should handle a click event', () => {
    const props = {
      datum: {
        dataSets: [
          {
            stroke: '#000000',
            id: 'loremIpsum',
            isThreshold: false,
            data: [{ y: 0, hasData: true }]
          },
          {
            stroke: '#ff0000',
            id: 'dolorSit',
            isThreshold: true,
            data: [{ y: 0, isInfinite: false }]
          }
        ]
      },
      legend: {
        'test-dolorSit': true
      },
      productLabel: 'test',
      viewId: 'test'
    };

    const component = shallow(<GraphCardChartLegend {...props} />);
    expect(component.find(Button).first()).toMatchSnapshot('click event pre');

    component.find(Button).first().simulate('click');
    // emulate a Redux state update.
    component.setProps({
      legend: { ...props.legend, ...{ 'test-loremIpsum': true } }
    });
    expect(component.find(Button).first()).toMatchSnapshot('click event update');

    component.find(Button).first().simulate('keyPress');
    // emulate a Redux state update.
    component.setProps({
      legend: { ...props.legend, ...{ 'test-loremIpsum': false } }
    });
    expect(component.find(Button).first()).toMatchSnapshot('click event post');
  });

  it('should handle variations in data when returning legend items', () => {
    const props = {
      datum: {
        dataSets: [
          {
            stroke: '#000000',
            id: 'loremIpsum',
            isThreshold: false,
            data: [{ y: 0, hasData: true }]
          },
          {
            stroke: '#ff0000',
            id: 'dolorSit',
            isThreshold: true,
            data: [{ y: 0, isInfinite: false }]
          }
        ]
      },
      productLabel: 'test'
    };

    const component = shallow(<GraphCardChartLegend {...props} />);

    expect(
      component.instance().renderLegendItem({
        chartId: 'loremIpsum',
        color: '#000000',
        isDisabled: false,
        isThreshold: false,
        labelContent: 'lorem ispum',
        tooltipContent: 'dolor sit'
      })
    ).toMatchSnapshot('legend item, WITH tooltip content');

    expect(
      component.instance().renderLegendItem({
        chartId: 'loremIpsum',
        color: '#000000',
        isDisabled: false,
        isThreshold: false,
        labelContent: 'lorem ispum',
        tooltipContent: undefined
      })
    ).toMatchSnapshot('legend item, MISSING tooltip content');

    expect(
      component.instance().renderLegendItem({
        chartId: 'loremIpsum',
        color: '#000000',
        isDisabled: true,
        isThreshold: false,
        labelContent: 'lorem ispum',
        tooltipContent: undefined
      })
    ).toMatchSnapshot('legend item, disabled');
  });
});
