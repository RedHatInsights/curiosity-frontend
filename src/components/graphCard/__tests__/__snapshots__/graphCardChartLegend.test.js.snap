// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphCardChartLegend Component should handle a click event: click event post 1`] = `
<Component
  className="victory-legend-item"
  component="a"
  icon={
    <div
      aria-hidden={true}
      className="legend-icon"
      style={
        Object {
          "backgroundColor": "#000000",
          "visibility": "visible",
        }
      }
    />
  }
  isDisabled={false}
  key="curiosity-button-loremIpsum"
  onClick={[Function]}
  onKeyPress={[Function]}
  tabIndex={0}
  variant="link"
>
  t(curiosity-graph.loremIpsumLabel,curiosity-graph.noLabel, [object Object])
</Component>
`;

exports[`GraphCardChartLegend Component should handle a click event: click event pre 1`] = `
<Component
  className="victory-legend-item"
  component="a"
  icon={
    <div
      aria-hidden={true}
      className="legend-icon"
      style={
        Object {
          "backgroundColor": "#000000",
          "visibility": "visible",
        }
      }
    />
  }
  isDisabled={false}
  key="curiosity-button-loremIpsum"
  onClick={[Function]}
  onKeyPress={[Function]}
  tabIndex={0}
  variant="link"
>
  t(curiosity-graph.loremIpsumLabel,curiosity-graph.noLabel, [object Object])
</Component>
`;

exports[`GraphCardChartLegend Component should handle a click event: click event update 1`] = `
<Component
  className="victory-legend-item"
  component="a"
  icon={
    <EyeSlashIcon
      color="currentColor"
      noVerticalAlign={false}
      size="sm"
      title={null}
    />
  }
  isDisabled={false}
  key="curiosity-button-loremIpsum"
  onClick={[Function]}
  onKeyPress={[Function]}
  tabIndex={0}
  variant="link"
>
  t(curiosity-graph.loremIpsumLabel,curiosity-graph.noLabel, [object Object])
</Component>
`;

exports[`GraphCardChartLegend Component should handle variations in data when returning legend items: legend item, MISSING tooltip content 1`] = `
<Unknown
  className="victory-legend-item"
  component="a"
  icon={
    <div
      aria-hidden={true}
      className="legend-icon"
      style={
        Object {
          "backgroundColor": "#000000",
          "visibility": "visible",
        }
      }
    />
  }
  isDisabled={false}
  onClick={[Function]}
  onKeyPress={[Function]}
  tabIndex={0}
  variant="link"
>
  lorem ispum
</Unknown>
`;

exports[`GraphCardChartLegend Component should handle variations in data when returning legend items: legend item, WITH tooltip content 1`] = `
<Tooltip
  appendTo={[Function]}
  aria="describedby"
  boundary="window"
  className=""
  content={
    <p>
      dolor sit
    </p>
  }
  distance={-10}
  enableFlip={true}
  entryDelay={100}
  exitDelay={0}
  flipBehavior={
    Array [
      "top",
      "right",
      "bottom",
      "left",
      "top",
      "right",
      "bottom",
    ]
  }
  id=""
  isAppLauncher={false}
  isContentLeftAligned={false}
  isVisible={false}
  maxWidth="18.75rem"
  position="top"
  tippyProps={Object {}}
  trigger="mouseenter focus"
  zIndex={9999}
>
  <Unknown
    className="victory-legend-item"
    component="a"
    icon={
      <div
        aria-hidden={true}
        className="legend-icon"
        style={
          Object {
            "backgroundColor": "#000000",
            "visibility": "visible",
          }
        }
      />
    }
    isDisabled={false}
    onClick={[Function]}
    onKeyPress={[Function]}
    tabIndex={0}
    variant="link"
  >
    lorem ispum
  </Unknown>
</Tooltip>
`;

exports[`GraphCardChartLegend Component should handle variations in data when returning legend items: legend item, disabled 1`] = `
<Unknown
  className="victory-legend-item"
  component="a"
  icon={
    <EyeSlashIcon
      color="currentColor"
      noVerticalAlign={false}
      size="sm"
      title={null}
    />
  }
  isDisabled={true}
  onClick={[Function]}
  onKeyPress={[Function]}
  tabIndex={0}
  variant="link"
>
  lorem ispum
</Unknown>
`;

exports[`GraphCardChartLegend Component should render a basic component: basic 1`] = `
<Fragment>
  <Tooltip
    appendTo={[Function]}
    aria="describedby"
    boundary="window"
    className=""
    content={
      <p>
        t(curiosity-graph.loremIpsumLegendTooltip, [object Object])
      </p>
    }
    distance={-10}
    enableFlip={true}
    entryDelay={100}
    exitDelay={0}
    flipBehavior={
      Array [
        "top",
        "right",
        "bottom",
        "left",
        "top",
        "right",
        "bottom",
      ]
    }
    id=""
    isAppLauncher={false}
    isContentLeftAligned={false}
    isVisible={false}
    key="curiosity-tooltip-loremIpsum"
    maxWidth="18.75rem"
    position="top"
    tippyProps={Object {}}
    trigger="mouseenter focus"
    zIndex={9999}
  >
    <Component
      className="victory-legend-item"
      component="a"
      icon={
        <div
          aria-hidden={true}
          className="legend-icon"
          style={
            Object {
              "backgroundColor": "#000000",
              "visibility": "visible",
            }
          }
        />
      }
      isDisabled={false}
      key="curiosity-button-loremIpsum"
      onClick={[Function]}
      onKeyPress={[Function]}
      tabIndex={0}
      variant="link"
    >
      t(curiosity-graph.loremIpsumLabel,curiosity-graph.noLabel, [object Object])
    </Component>
  </Tooltip>
</Fragment>
`;

exports[`GraphCardChartLegend Component should render basic data: data 1`] = `
<Fragment>
  <Tooltip
    appendTo={[Function]}
    aria="describedby"
    boundary="window"
    className=""
    content={
      <p>
        t(curiosity-graph.loremIpsumLegendTooltip, [object Object])
      </p>
    }
    distance={-10}
    enableFlip={true}
    entryDelay={100}
    exitDelay={0}
    flipBehavior={
      Array [
        "top",
        "right",
        "bottom",
        "left",
        "top",
        "right",
        "bottom",
      ]
    }
    id=""
    isAppLauncher={false}
    isContentLeftAligned={false}
    isVisible={false}
    key="curiosity-tooltip-loremIpsum"
    maxWidth="18.75rem"
    position="top"
    tippyProps={Object {}}
    trigger="mouseenter focus"
    zIndex={9999}
  >
    <Component
      className="victory-legend-item"
      component="a"
      icon={
        <div
          aria-hidden={true}
          className="legend-icon"
          style={
            Object {
              "backgroundColor": "#000000",
              "visibility": "visible",
            }
          }
        />
      }
      isDisabled={false}
      key="curiosity-button-loremIpsum"
      onClick={[Function]}
      onKeyPress={[Function]}
      tabIndex={0}
      variant="link"
    >
      t(curiosity-graph.loremIpsumLabel,curiosity-graph.noLabel, [object Object])
    </Component>
  </Tooltip>
  <Tooltip
    appendTo={[Function]}
    aria="describedby"
    boundary="window"
    className=""
    content={
      <p>
        t(curiosity-graph.ametConsecteturLegendTooltip, [object Object])
      </p>
    }
    distance={-10}
    enableFlip={true}
    entryDelay={100}
    exitDelay={0}
    flipBehavior={
      Array [
        "top",
        "right",
        "bottom",
        "left",
        "top",
        "right",
        "bottom",
      ]
    }
    id=""
    isAppLauncher={false}
    isContentLeftAligned={false}
    isVisible={false}
    key="curiosity-tooltip-ametConsectetur"
    maxWidth="18.75rem"
    position="top"
    tippyProps={Object {}}
    trigger="mouseenter focus"
    zIndex={9999}
  >
    <Component
      className="victory-legend-item"
      component="a"
      icon={
        <EyeSlashIcon
          color="currentColor"
          noVerticalAlign={false}
          size="sm"
          title={null}
        />
      }
      isDisabled={true}
      key="curiosity-button-ametConsectetur"
      onClick={[Function]}
      onKeyPress={[Function]}
      tabIndex={0}
      variant="link"
    >
      t(curiosity-graph.ametConsecteturLabel,curiosity-graph.noLabel, [object Object])
    </Component>
  </Tooltip>
  <Tooltip
    appendTo={[Function]}
    aria="describedby"
    boundary="window"
    className=""
    content={
      <p>
        t(curiosity-graph.dolorSitLegendTooltip,curiosity-graph.thresholdLegendTooltip, [object Object])
      </p>
    }
    distance={-10}
    enableFlip={true}
    entryDelay={100}
    exitDelay={0}
    flipBehavior={
      Array [
        "top",
        "right",
        "bottom",
        "left",
        "top",
        "right",
        "bottom",
      ]
    }
    id=""
    isAppLauncher={false}
    isContentLeftAligned={false}
    isVisible={false}
    key="curiosity-tooltip-dolorSit"
    maxWidth="18.75rem"
    position="top"
    tippyProps={Object {}}
    trigger="mouseenter focus"
    zIndex={9999}
  >
    <Component
      className="victory-legend-item"
      component="a"
      icon={
        <hr
          aria-hidden={true}
          className="threshold-legend-icon"
          style={
            Object {
              "borderTopColor": "#ff0000",
              "visibility": "visible",
            }
          }
        />
      }
      isDisabled={false}
      key="curiosity-button-dolorSit"
      onClick={[Function]}
      onKeyPress={[Function]}
      tabIndex={0}
      variant="link"
    >
      t(curiosity-graph.dolorSitLabel,curiosity-graph.thresholdLabel, [object Object])
    </Component>
  </Tooltip>
  <Tooltip
    appendTo={[Function]}
    aria="describedby"
    boundary="window"
    className=""
    content={
      <p>
        t(curiosity-graph.nonCursusLegendTooltip,curiosity-graph.thresholdLegendTooltip, [object Object])
      </p>
    }
    distance={-10}
    enableFlip={true}
    entryDelay={100}
    exitDelay={0}
    flipBehavior={
      Array [
        "top",
        "right",
        "bottom",
        "left",
        "top",
        "right",
        "bottom",
      ]
    }
    id=""
    isAppLauncher={false}
    isContentLeftAligned={false}
    isVisible={false}
    key="curiosity-tooltip-nonCursus"
    maxWidth="18.75rem"
    position="top"
    tippyProps={Object {}}
    trigger="mouseenter focus"
    zIndex={9999}
  >
    <Component
      className="victory-legend-item"
      component="a"
      icon={
        <hr
          aria-hidden={true}
          className="threshold-legend-icon"
          style={
            Object {
              "borderTopColor": "#ff0000",
              "visibility": "visible",
            }
          }
        />
      }
      isDisabled={false}
      key="curiosity-button-nonCursus"
      onClick={[Function]}
      onKeyPress={[Function]}
      tabIndex={0}
      variant="link"
    >
      t(curiosity-graph.nonCursusLabel,curiosity-graph.thresholdLabel, [object Object])
    </Component>
  </Tooltip>
</Fragment>
`;
